### PR TITLE
[21229] Improve CI workflows in master

### PR DESCRIPTION
**Pull request by JesusPoderoso**
## Description

This PR improves the GitHub CI of this repo, to make it as similar as the one in Fast DDS

**Note**: While backporting this PR, please ensure the following changes:
- Remove nightly jobs
- Update `ubuntu-ci.yml` and `windows-ci.yml` files with the proper fallback branches
- Change `fastdds.repos` to `fastrtps.repos`
- 2.x branches require both `v141` and `v142` vs toolsets
- 2.6.x requires building on `ubuntu-20.04`

For those reasons, a **manual backport to 2.14.x** will be opened, and that one will be backported to remain branches.

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- **N/A** Changes do not break current interoperability.
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- **N/A** Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] CI passes without warnings or errors.

### CODE REVIEW

File: .github/workflows/reusable-windows-ci.yml [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        type: boolean
+        default: true
+env:
+  colcon-build-default-cmake-args: '-DTHIRDPARTY_Asio=FORCE -DTHIRDPARTY_TinyXML2=FORCE -DTHIRDPARTY_fastcdr=OFF -DTHIRDPARTY_UPDATE=ON -DEPROSIMA_EXTRA_CMAKE_CXX_FLAGS="/MP /WX"'
+defaults:
+  run:
+    shell: pwsh
+
+jobs:
+  shapes_demo_build:
```
Comment (elianalf) [inline .github/workflows/reusable-windows-ci.yml:52]:
```suggestion
  reusable-windows-ci:
```

File: .github/workflows/reusable-ubuntu-ci.yml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        type: boolean
+        default: false
+env:
+    colcon-build-default-cmake-args: '-DCMAKE_CXX_FLAGS="-Werror -Wall -Wextra -Wpedantic -Wunused-value -Woverloaded-virtual"'
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  shapes_demo_build:
```
Comment (elianalf) [inline .github/workflows/reusable-ubuntu-ci.yml:52]:
```suggestion
  reusable-ubuntu-ci:
```

File: .github/workflows/ubuntu-ci.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
+
```
Comment (elianalf) [inline .github/workflows/ubuntu-ci.yml:1]:
```suggestion
```

File: .github/workflows/ubuntu-ci.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      - '**.md'
+      - '**.txt'
+      - '!**/CMakeLists.txt'
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  ubuntu-ci:
```
Comment (elianalf) [inline .github/workflows/ubuntu-ci.yml:51]: I think that this check can be moved here, so when a PR has conflicts we can see the jobs skipped
```suggestion
  ubuntu-ci:
        if: ${{ (
              !contains(github.event.pull_request.labels.*.name, 'conflicts')
            ) }}
```

File: .github/workflows/reusable-windows-ci.yml (hunk 1 above)
Comment (elianalf) [inline .github/workflows/reusable-windows-ci.yml:52]:
```suggestion
  reusable-windows-ci:
```

File: .github/workflows/reusable-ubuntu-ci.yml (hunk 2 above)
Comment (elianalf) [inline .github/workflows/reusable-ubuntu-ci.yml:52]:
```suggestion
  reusable-ubuntu-ci:
```

**Review by elianalf — CHANGES_REQUESTED** [review]

File: .github/workflows/nightly-windows-ci.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+          - 'v141'
+          - 'v142'
+    uses: eProsima/ShapesDemo/.github/workflows/reusable-windows-ci.yml@2.6.x
+    with:
+      os-version: ${{ matrix.os-version }}
+      vs.toolset: ${{ matrix.vs-toolset }}
+      label: 'nightly-${{ matrix.os-version }}-${{ matrix.vs-toolset }}-ci-2.6.x'
+      shapes-demo-branch: '2.6.x'
+      fastdds-branch: '2.6.x'
+      run-build: false
```
[the remark this answers is not in the document]
Comment (JesusPoderoso) [reply .github/workflows/nightly-windows-ci.yml:96]: 🤦 ups

File: .github/workflows/reusable-windows-ci.yml [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      shapes-demo-branch:
+        description: 'Branch or tag of Shapes Demo repository'
+        required: true
+        type: string
+      fastdds-branch:
+        description: 'Branch or tag of Fast DDS repository'
+        required: true
+        type: string
+      run-build:
+        description: 'Build Fast DDS Python (CI skipped otherwise)'
```
[the remark this answers is not in the document]
Comment (JesusPoderoso) [reply .github/workflows/reusable-windows-ci.yml:41]: 🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦v

**Comment by JesusPoderoso** [comment]
@Mergifyio backport 2.14.x 2.13.x 2.10.x 2.6.x

**Pull request merged**